File: mm-cs/src/MetaMessage/Mm/MmTree.cs
namespace MetaMessage.Mm;

public interface IMmTree
{
    MmTag Tag { get; }
}

public class MmScalar : IMmTree
{
    public object Data { get; set; }
    public string Text { get; set; }
    public MmTag Tag { get; set; }

    public MmScalar(object data, string text, MmTag tag)
    {
        Data = data;
        Text = text;
        Tag = tag;
    }
}

public class MmArray : IMmTree
{
    public List<IMmTree> Children { get; set; }
    public MmTag Tag { get; set; }

    public MmArray(List<IMmTree> children, MmTag tag)
    {
        Children = children;
        Tag = tag;
    }
}

public class MmMap : IMmTree
{
    public List<KeyValuePair<MmScalar, IMmTree>> Entries { get; set; }
    public MmTag Tag { get; set; }

    public MmMap(List<KeyValuePair<MmScalar, IMmTree>> entries, MmTag tag)
    {
        Entries = entries;
        Tag = tag;
    }
}

File: mm-php/src/io/metamessage/mm/MmTree.php
<?php

namespace io\metamessage\mm;

class MmTree {
    public MmTag $tag;

    public function __construct(MmTag $tag) {
        $this->tag = $tag;
    }
}

namespace io\metamessage\mm\MmTree;

use io\metamessage\mm\MmTag;

class MmScalar extends \io\metamessage\mm\MmTree {
    public $data;
    public string $text;

    public function __construct($data, string $text, MmTag $tag) {
        parent::__construct($tag);
        $this->data = $data;
        $this->text = $text;
    }
}

class MmObject extends \io\metamessage\mm\MmTree {
    public array $fields;

    public function __construct(MmTag $tag, array $fields) {
        parent::__construct($tag);
        $this->fields = $fields;
    }
}

class MmArray extends \io\metamessage\mm\MmTree {
    public array $items;

    public function __construct(MmTag $tag, array $items) {
        parent::__construct($tag);
        $this->items = $items;
    }
}
